## **1. Importações**

In [2]:
import pandas as pd


## **2. Leitura dos Dados**

São lidas 8 fontes de dados heterogéneas, cada uma com frequência, formato de data e estrutura de colunas distintos. A EA-MD-QD (mensal e trimestral) serve de âncora estrutural do dataset — começa em Jan 2000 e todas as restantes fontes são integradas sobre ela via left join.

| Fonte | Descrição | Período |
|---|---|---|
| `target` | [Inflação portuguesa (IPC, taxa de variação homóloga)](https://bpstat.bportugal.pt/serie/5721524) | 1949–2025 |
| `ea-md` | [Indicadores macroeconómicos mensais da Área do Euro para PT](https://zenodo.org/records/18804061) | 2000–2026 |
| `ea-qd` | [Indicadores macroeconómicos trimestrais da Área do Euro para PT](https://zenodo.org/records/18804061) | 2000–2026 |
| `fred-md` | [Indicadores económicos e financeiros dos EUA (FRED-MD)](https://www.stlouisfed.org/research/economists/mccracken/fred-databases) | 1959–2025 |
| `epu` | [Índice de Incerteza de Política Económica — Portugal](https://github.com/hmorao95/epu_portugal/tree/main) | 1998–2025 |
| `unemployment` | [Taxa de desemprego mensal em Portugal](https://data-explorer.oecd.org/vis?df[ds]=dsDisseminateFinalDMZ&df[id]=DSD_LFS%40DF_IALFS_UNE_M&df[ag]=OECD.SDD.TPS&df[vs]=1.0&pd=%2C&dq=PRT..._Z.Y._T.Y_GE15..M&ly[cl]=TIME_PERIOD&to[TIME_PERIOD]=false&vw=tb) | 1983–2025 |
| `exchange` | [Taxa de câmbio EUR/USD (fim de mês)](https://bpstat.bportugal.pt/serie/12532123) | 1999–2026 |
| `ppi` | [Índice de Preços no Produtor — Portugal (taxa de variação homóloga)](https://www.econdb.com/series/PPIPT~TOYA/portugal-producer-price-index/?from=1999-10-11&to=2026-04-20) | 2000–2026 |

In [24]:
df_target = pd.read_csv("dados/target.csv", encoding='"latin1"', sep=';')
df_fred_md = pd.read_csv("dados/2026-01-MD.csv")
df_ppi = pd.read_csv("dados/PPI_PT.csv")

df_epu = pd.read_excel("dados/epuptindex_data.xlsx")
df_unemployment = pd.read_excel("dados/monthly unemployment rates.xlsx", skiprows=7)

df_ea_md = pd.read_excel("dados/EA-MD-QD-2026-02/data_TR3/monthly-PT.xlsx")
df_ea_qd = pd.read_excel("dados/EA-MD-QD-2026-02/data_TR3/quarter-PT.xlsx")

df_exchange = pd.read_csv("dados/exchange_rate.csv", encoding='"latin1"', sep=';')

In [25]:
df_target.head() # Apenas o Periodo de Referência e o Valor são relevantes

,# Série,Designação da série (PT),Métrica,Unidade de Medida,Período de referência,Valor,Estado
0,5721524,IPC total-taxa de variação homóloga,Taxa de variação homóloga,Percentagem,31/01/1949,4.4,NaN
1,5721524,IPC total-taxa de variação homóloga,Taxa de variação homóloga,Percentagem,28/02/1949,4.2,NaN
2,5721524,IPC total-taxa de variação homóloga,Taxa de variação homóloga,Percentagem,31/03/1949,4.8,NaN
3,5721524,IPC total-taxa de variação homóloga,Taxa de variação homóloga,Percentagem,30/04/1949,3.8,NaN
4,5721524,IPC total-taxa de variação homóloga,Taxa de variação homóloga,Percentagem,31/05/1949,4.5,NaN


In [26]:
df_ea_md.head()

,Time,UNETOT_PT,UNEO25_PT,UNEU25_PT,REER42_PT,LTIRT_PT,IPMN_PT,IPCAG_PT,IPCOG_PT,IPDCOG_PT,...,HICPNG_PT,ICONFIX_PT,CCONFIX_PT,ESENTIX_PT,KCONFIX_PT,RTCONFIX_PT,SCONFIX_PT,BCI_PT,CCI_PT,SHIX_PT
0,2000-01-01,5.6,4.4,12.7,96.14621,5.81,123.5,191.3,118.0,129.6,...,42.21,7.3,-2.4,113.9,-15.8,9.4,10.3,102.2262,102.3880,110.4053
1,2000-02-01,5.4,4.1,13.0,95.12462,5.78,122.7,177.4,119.8,131.1,...,42.29,8.9,-3.0,116.7,-5.6,11.0,14.2,102.5146,102.2470,124.9385
2,2000-03-01,5.3,4.0,12.4,94.49293,5.61,116.5,164.6,116.5,124.8,...,42.41,7.7,-3.2,116.2,-0.1,11.7,11.9,102.6476,101.9521,128.3406
3,2000-04-01,5.2,3.9,12.8,94.90064,5.52,115.5,159.3,118.0,125.3,...,44.71,8.8,-9.2,114.1,-13.5,6.1,13.2,102.7988,101.5175,115.6546
4,2000-05-01,5.3,4.0,13.1,94.96760,5.68,120.8,171.0,121.0,127.9,...,44.85,10.1,-8.0,113.3,-3.7,1.5,6.0,102.8972,101.2843,112.0247


In [27]:
df_ea_qd.head()

,Time,GDP_PT,EXPGS_PT,IMPGS_PT,GFCE_PT,HFCE_PT,CONSD_PT,GCF_PT,GFCF_PT,GFACON_PT,...,HHLB.LLN_PT,ULCIN_PT,ULCMQ_PT,ULCMN_PT,ULCCON_PT,ULCRT_PT,ULCFC_PT,ULCRE_PT,ULCPR_PT,DFGDP_PT
0,2000-01-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2000-02-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2000-03-01,43450.6,10151.5,13257.6,7387.6,28123.5,3524.1,10981.3,11097.5,7270.5,...,59309.7,64.8,NaN,NaN,NaN,72.2,NaN,NaN,NaN,72.43
3,2000-04-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2000-05-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [28]:
df_ppi.head()

,Date,PPIPT
0,2000-01-01,21.5
1,2000-02-01,21.9
2,2000-03-01,21.2
3,2000-04-01,20.0
4,2000-05-01,17.9


In [29]:
df_fred_md.head() # Eliminar a linha 0

,sasdate,RPI,W875RX1,DPCERA3M086SBEA,CMRMTSPLx,RETAILx,INDPRO,IPFPNSS,IPFINAL,IPCONGD,...,DNDGRG3M086SBEA,DSERRG3M086SBEA,CES0600000008,CES2000000008,CES3000000008,UMCSENTx,DTCOLNVHFNM,DTCTHFNM,INVEST,VIXCLSx
0,Transform:,5.000,5.0,5.000,5.0000,5.00000,5.0000,5.0000,5.0000,5.0000,...,6.000,6.000,6.00,6.00,6.00,2.0,6.0,6.0,6.0000,1.0
1,1/1/1959,2583.560,2426.0,15.188,276676.8154,17689.23968,21.9998,23.6312,22.5507,32.1377,...,18.294,10.152,2.13,2.45,2.04,NaN,6476.0,12298.0,84.2043,NaN
2,2/1/1959,2593.596,2434.8,15.346,278713.9773,17819.01912,22.4306,23.9501,22.7461,32.3734,...,18.302,10.167,2.14,2.46,2.05,NaN,6476.0,12298.0,83.5280,NaN
3,3/1/1959,2610.396,2452.7,15.491,277775.2539,17967.91336,22.7538,24.0951,22.8577,32.3734,...,18.289,10.185,2.15,2.45,2.07,NaN,6508.0,12349.0,81.6405,NaN
4,4/1/1959,2627.446,2470.0,15.435,283362.7075,17978.97983,23.2384,24.4431,23.1926,32.8841,...,18.300,10.221,2.16,2.47,2.08,NaN,6620.0,12484.0,81.8099,NaN


In [30]:
df_epu.head() # Eliminar as duas últimas colunas

,Unnamed: 0,epu_pt,Unnamed: 2,Cite it as follows:
0,1998m1,134.655810,NaN,"Morão, H. (2024). An economic policy uncertain..."
1,1998m2,91.742596,NaN,Regular updates:
2,1998m3,92.135717,NaN,https://github.com/hmorao95/epu_portugal
3,1998m4,112.026194,NaN,NaN
4,1998m5,125.879406,NaN,NaN


In [31]:
df_unemployment.head() # Eliminar a última coluna

,Time period,Value,Unnamed: 2
0,1983-Jan,8.1,NaN
1,1983-Feb,8.1,NaN
2,1983-Mar,8.3,NaN
3,1983-Apr,8.5,NaN
4,1983-May,8.8,NaN


In [32]:
df_exchange.head() # Apenas o Periodo de Referência e o Valor são relevantes

,# Série,Designação da série (PT),Métrica,Unidade de Medida,Período de referência,Valor,Estado
0,12532123,Dólar dos EUA (USD) - fim de mês,Valor em fim do período,Divisa,1999-01-31,1.1384,NaN
1,12532123,Dólar dos EUA (USD) - fim de mês,Valor em fim do período,Divisa,1999-02-28,1.1018,NaN
2,12532123,Dólar dos EUA (USD) - fim de mês,Valor em fim do período,Divisa,1999-03-31,1.0742,NaN
3,12532123,Dólar dos EUA (USD) - fim de mês,Valor em fim do período,Divisa,1999-04-30,1.0597,NaN
4,12532123,Dólar dos EUA (USD) - fim de mês,Valor em fim do período,Divisa,1999-05-31,1.0456,NaN


## **3. Pré-processamento**

### Formatos de Data por Fonte

- **target**: `Período de referência` — DD/MM/AAAA
- **fred-md**: `sasdate` — M/D/AAAA
- **epu**: `Unnamed: 0` — AAAAmM
- **unemployment**: `Time period` — AAAA-Mês
- **ea-md & ea-qd**: `Time` — AAAA-MM-DD
- **exchange**: `Período de referência` — AAAA-MM-DD
- **ppi**: `Date` — AAAA-MM-DD

Todas as datas são normalizadas para o primeiro dia do mês (`AAAA-MM-01`).

### **3.1 Seleção de Colunas**

Da FRED-MD são retidas apenas 15 colunas relevantes (produção industrial, mercado de trabalho, taxas de juro, preços e mercados financeiros) identificadas nas fontes de estudo. Das restantes fontes eliminam-se colunas auxiliares de metadados — ficam apenas a série temporal e a coluna de data.

In [33]:
df_target = df_target[['Período de referência', 'Valor']]

df_exchange = df_exchange[['Período de referência', 'Valor']]

cols_fred_global = [
    'sasdate', 'INDPRO', 'IPFINAL', 'IPCONGD',
    'AWOTMAN', 'AWHMAN', 'FEDFUNDS', 'GS1', 'GS5',
    'GS10', 'AAA', 'BAA', 'OILPRICEx', 'CUSR0000SAC',
    'PCEPI', 'S&P 500'
]
df_fred_md = df_fred_md.iloc[1:].reset_index(drop=True)[cols_fred_global]

df_epu.drop(df_epu.columns[-2:], axis=1, inplace=True)

df_unemployment.drop(df_unemployment.columns[-1], axis=1, inplace=True)

# df_ea_md # Está good
# df_ea_qd # Irá ser feita Interpolação e Auto-Arima

### **3.2 Normalização das Datas**

Todas as colunas de data são renomeadas para `Date`, os formatos heterogéneos são convertidos para `datetime` e os valores normalizados para o primeiro dia de cada mês — convenção adoptada em todo o projeto.

In [34]:
default_date = 'Date'

df_target.rename({'Período de referência': default_date}, axis=1, inplace=True)
df_fred_md.rename({'sasdate': default_date}, axis=1, inplace=True)
df_epu.rename({'Unnamed: 0': default_date}, axis=1, inplace=True)
df_unemployment.rename({'Time period': default_date}, axis=1, inplace=True)

df_ea_md.rename({'Time': default_date}, axis=1, inplace=True)
df_ea_qd.rename({'Time': default_date}, axis=1, inplace=True)

df_exchange.rename({'Período de referência': default_date}, axis=1, inplace=True)

In [35]:
df_unemployment.dropna(axis=0, inplace=True)
df_unemployment.reset_index(drop=True, inplace=True)

In [36]:
df_target['Date'] = pd.to_datetime(
    df_target['Date'],
    dayfirst=True
)

df_fred_md['Date'] = pd.to_datetime(
    df_fred_md['Date']
)

df_ea_md['Date'] = pd.to_datetime(
    df_ea_md['Date']
)

df_ea_qd['Date'] = pd.to_datetime(
    df_ea_qd['Date']
)

df_ppi['Date'] = pd.to_datetime(
    df_ppi['Date']
)

df_epu['Date'] = pd.to_datetime(
    df_epu['Date'].str.replace('m', '/'),
    format='%Y/%m'
)

df_unemployment['Date'] = pd.to_datetime(
    df_unemployment['Date'],
    format='%Y-%b'
)

In [37]:
df_exchange['Date'] = pd.to_datetime(
    df_exchange['Date'],
    format='%Y-%m-%d'
)

In [38]:
df_target['Date'] = df_target.Date.dt.to_period("M").dt.to_timestamp()

df_exchange['Date'] = df_exchange.Date.dt.to_period("M").dt.to_timestamp()

In [39]:
dfs = {
    'target': df_target, 'fred-md': df_fred_md,
    'ea-md': df_ea_md, 'ea-qd': df_ea_qd,
    'epu': df_epu, 'unemployment': df_unemployment,
    'exchange': df_exchange, 'ppi': df_ppi
}

for df_name, df in dfs.items():
    print(f"Data mínima de {df_name} é `{df['Date'].min()}`")

Data mínima de target é `1949-01-01 00:00:00`
Data mínima de fred-md é `1959-01-01 00:00:00`
Data mínima de ea-md é `2000-01-01 00:00:00`
Data mínima de ea-qd é `2000-01-01 00:00:00`
Data mínima de epu é `1998-01-01 00:00:00`
Data mínima de unemployment é `1983-01-01 00:00:00`
Data mínima de exchange é `1999-01-01 00:00:00`
Data mínima de ppi é `2000-01-01 00:00:00`


In [40]:
for df_name, df in dfs.items():
    print(f"Data máxima de {df_name} é `{df['Date'].max()}`")

Data máxima de target é `2025-12-01 00:00:00`
Data máxima de fred-md é `2025-12-01 00:00:00`
Data máxima de ea-md é `2026-02-01 00:00:00`
Data máxima de ea-qd é `2026-02-01 00:00:00`
Data máxima de epu é `2025-10-01 00:00:00`
Data máxima de unemployment é `2025-12-01 00:00:00`
Data máxima de exchange é `2026-02-01 00:00:00`
Data máxima de ppi é `2026-02-01 00:00:00`


In [44]:
total = 0
for df_name, df in dfs.items():
    print(f"{df_name} dataset: {df.shape[1] - 1} coluna/s")
    total += df.shape[1] - 1 # Eliminar data

print("=" * 30)
print(f"Total: {total} colunas")

target dataset: 1 coluna/s
fred-md dataset: 15 coluna/s
ea-md dataset: 34 coluna/s
ea-qd dataset: 65 coluna/s
epu dataset: 1 coluna/s
unemployment dataset: 1 coluna/s
exchange dataset: 1 coluna/s
ppi dataset: 1 coluna/s
Total: 119 colunas


## **4. Integração dos Dados**

As colunas de cada fonte são renomeadas com o sufixo da origem (e.g. `INDPRO_fred-md`) para evitar colisões. O merge é feito por left join sobre a EA-MD-QD mensal. O dataset resultante é truncado a Out 2025, data da última observação para o indice de incerteza politica.

In [45]:
df_target.rename({'Valor': 'inflation'}, axis=1, inplace=True)
df_unemployment.rename({'Value': 'unemployment_up_15'}, axis=1, inplace=True)
df_exchange.rename({'Valor': 'exchange'}, axis=1, inplace=True)

In [46]:
for df_name, df in dfs.items():
    rename = dict()
    for col in df.columns[1:]:
        rename[col] = f'{col}_{df_name}'
    df.rename(rename, axis=1, inplace=True)

In [47]:
df = pd.merge(df_ea_md, df_ea_qd, 'left', 'Date')
df = pd.merge(df, df_fred_md, 'left', 'Date')
df = pd.merge(df, df_epu, 'left', 'Date')
df = pd.merge(df, df_unemployment, 'left', 'Date')
df = pd.merge(df, df_exchange, 'left', 'Date')
df = pd.merge(df, df_ppi, 'left', 'Date')
df = pd.merge(df, df_target, 'left', 'Date')
df.head()

,Date,UNETOT_PT_ea-md,UNEO25_PT_ea-md,UNEU25_PT_ea-md,REER42_PT_ea-md,LTIRT_PT_ea-md,IPMN_PT_ea-md,IPCAG_PT_ea-md,IPCOG_PT_ea-md,IPDCOG_PT_ea-md,...,BAA_fred-md,OILPRICEx_fred-md,CUSR0000SAC_fred-md,PCEPI_fred-md,S&P 500_fred-md,epu_pt_epu,unemployment_up_15_unemployment,exchange_exchange,PPIPT_ppi,inflation_target
0,2000-01-01,5.6,4.4,12.7,96.14621,5.81,123.5,191.3,118.0,129.6,...,8.33,27.18,146.7,72.961,1425.59,107.096201,5.6,0.9791,21.5,2.1
1,2000-02-01,5.4,4.1,13.0,95.12462,5.78,122.7,177.4,119.8,131.1,...,8.29,29.35,147.6,73.191,1388.87,105.539878,5.4,0.9714,21.9,1.8
2,2000-03-01,5.3,4.0,12.4,94.49293,5.61,116.5,164.6,116.5,124.8,...,8.37,29.89,149.1,73.505,1442.21,68.844150,5.3,0.9553,21.2,1.5
3,2000-04-01,5.2,3.9,12.8,94.90064,5.52,115.5,159.3,118.0,125.3,...,8.40,25.74,148.5,73.444,1461.36,70.292778,5.2,0.9085,20.0,2.1
4,2000-05-01,5.3,4.0,13.1,94.96760,5.68,120.8,171.0,121.0,127.9,...,8.90,28.78,148.5,73.505,1418.48,78.451666,5.3,0.9303,17.9,2.5


In [48]:
df = df.loc[df.Date <= '2025-10-01']

## **5. Engenharia de Atributos**

O preço do petróleo da FRED-MD está em USD. Para torná-lo comparável com a economia da Área do Euro, converte-se para euros dividindo pela taxa de câmbio EUR/USD do mesmo mês.

In [49]:
df['OILPRICEx_fred-md_Eur'] = df['OILPRICEx_fred-md'] / df['exchange_exchange']

## **6. Exportação**

O dataset final contém 310 observações mensais (Jan 2000 – Out 2025) e 121 variáveis. É guardado em `dados/CompactedData.csv` e serve de input para o `DataProcessing.ipynb`.

In [ ]:
df.shape

(310, 121)

In [52]:
df.to_csv('dados/CompactedData.csv', index=False)